### ISSUE TITLE
Incorrect "unreachable pattern" warning in Rust 1.50.

### ISSUE THREAD

**Issue by AlephAlpha** [rust-lang/rust#82012]
This seems to be a new issue in Rust 1.50. The latest nightly does not have this issue.

```rust
macro_rules! some {
    ($item:pat) => {
        Some($item)
    };
}

fn main() {
    let a = Some('A');
    match a {
        some!('a') | some!('A') => println!("A!"),
        _ => println!("Not a!"),
    }
}
```

[playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=dcd7b79f01f56a9ebef59805c54543ff)

It prints `"A!"` without any error, but gives the following warning.

```
warning: unreachable pattern
  --> src/main.rs:3:9
   |
3  |         Some($item)
   |         ^^^^^^^^^^^
...
10 |         some!('a') | some!('A') => println!("A!"),
   |                      ---------- in this macro invocation
   |
   = note: `#[warn(unreachable_patterns)]` on by default
   = note: this warning originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)

warning: 1 warning emitted
```

`some!('A')` is obviously reachable, otherwise it won't print `"A!"`.

**Comment by jyn514** [comment]
This is https://github.com/rust-lang/rust/issues/80501. It was fixed in https://github.com/rust-lang/rust/pull/80632 but not backported to stable.

**Comment by jyn514** [comment]
cc @rust-lang/release, I don't know the policy for this sort of thing.

**Comment by Mark-Simulacrum** [comment]
Gah. The relevant PR was not tagged beta-nominated so was overlooked; as it closed https://github.com/rust-lang/rust/issues/80501 the release team did not see that the regression was unfixed.

I am not sure this is worth a stable point release over, but I've now nominated the PR for beta backport (to 1.51, as it landed in 1.52) and stable backport (to 1.50).

**Comment by workingjubilee** [comment]
Resolved by the sands of time.
